Some fixes to smooth_l1_loss

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45532 Some fixes to smooth_l1_loss**

- updated documentation
- explicitly not supporting negative values for beta (previously the
result was incorrect)
- Removing default value for beta in the backwards function, since it's
only used internally by autograd (as per convention)

Differential Revision: [D24002415](https://our.internmc.facebook.com/intern/diff/D24002415)

bdhirsh committed:
Some fixes to smooth_l1_loss
- updated documentation
- explicitly not supporting negative values for beta (previously the
result was incorrect)
- Removing default value for beta in the backwards function, since it's
only used internally by autograd (as per convention)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -296,7 +296,8 @@ Tensor soft_margin_loss(
 Tensor smooth_l1_loss(const Tensor& input, const Tensor& target, const int64_t reduction, double beta) {
-  if (beta <= 0)
+  TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
+  if (beta == 0)
```
nit: we usually surround with `{}` or put the call on the same line.  It's too easy to assume indent scope means something.

gchanan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -305,7 +306,8 @@ Tensor smooth_l1_loss(const Tensor& input, const Tensor& target, const int64_t r
 Tensor& smooth_l1_loss_out(Tensor& result, const Tensor& input, const Tensor& target, int64_t reduction, double beta) {
-  if (beta <= 0)
+  TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
+  if (beta == 0)
```
same here.

gchanan commented on torch/nn/modules/loss.py:
```diff
@@ -778,7 +778,8 @@ class SmoothL1Loss(_Loss):
 
-    beta is an optional parameter that defaults to 1.
+    beta is an optional parameter that defaults to 1. When beta is set to 0, we call out directly
+    to l1_loss. Passing a negative value in for beta will result in an exception.
```
nit: you don't need to include implementation details here (call out directly to l1_loss).  You can just say something like, "Note: when beta is set to 0, this is equivalent to `:module:L1Loss` (not sure the formatting).

bdhirsh replied:
(I think it's :class: judging from some of the other code)

bdhirsh committed:
Update on "Some fixes to smooth_l1_loss"


- updated documentation
- explicitly not supporting negative values for beta (previously the
result was incorrect)
- Removing default value for beta in the backwards function, since it's
only used internally by autograd (as per convention)

Differential Revision: [D24002415](https://our.internmc.facebook.com/intern/diff/D24002415)

[ghstack-poisoned]
